# Pokemon Data Analysis
---

## Importing Required Modules 
1. importing numpy for mathematical operation on arrays and dataframe.
2. importing pandas for reading data and data manipualtion.

In [103]:
import numpy as np
import pandas as pd

## Reading dataaset

In [104]:
df=pd.read_csv('pokemon.csv')

In [105]:
df.head() ### viewing the data 

#                   Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1              Bulbasaur  Grass  Poison  45      49       49       65   
1  2                Ivysaur  Grass  Poison  60      62       63       80   
2  3               Venusaur  Grass  Poison  80      82       83      100   
3  3  VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4             Charmander   Fire     NaN  39      52       43       60   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False

In [4]:
df.shape

(800, 12)

## To read 10 rows from starting

In [106]:
df.head(10)

#                       Name Type 1  Type 2  HP  Attack  Defense  Sp. Atk  \
0  1                  Bulbasaur  Grass  Poison  45      49       49       65   
1  2                    Ivysaur  Grass  Poison  60      62       63       80   
2  3                   Venusaur  Grass  Poison  80      82       83      100   
3  3      VenusaurMega Venusaur  Grass  Poison  80     100      123      122   
4  4                 Charmander   Fire     NaN  39      52       43       60   
5  5                 Charmeleon   Fire     NaN  58      64       58       80   
6  6                  Charizard   Fire  Flying  78      84       78      109   
7  6  CharizardMega Charizard X   Fire  Dragon  78     130      111      130   
8  6  CharizardMega Charizard Y   Fire  Flying  78     104       78      159   
9  7                   Squirtle  Water     NaN  44      48       65       50   

   Sp. Def  Speed  Generation  Legendary  
0       65     45           1      False  
1       80     60           1      False  
2      100     80           1      False  
3      120     80           1      False  
4       50     65           1      False  
5       65     80           1      False  
6       85    100           1      False  
7       85    100           1      False  
8      115    100           1      False  
9       64     43           1      False

## To read last 5 rows

In [6]:
df.tail()

#                 Name   Type 1 Type 2  HP  Attack  Defense  Sp. Atk  \
795  719              Diancie     Rock  Fairy  50     100      150      100   
796  719  DiancieMega Diancie     Rock  Fairy  50     160      110      160   
797  720  HoopaHoopa Confined  Psychic  Ghost  80     110       60      150   
798  720   HoopaHoopa Unbound  Psychic   Dark  80     160       60      170   
799  721            Volcanion     Fire  Water  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True

### Short Description abot the Dataset
- Pokedex entry number of the Pokemon

pokedex is an  catalog that provide information regarding the various species of Pokémon featured in the Pokémon video game, anime and manga series.)
- Name = name of the Pokemon
- Type 1 = each Pokemon has a type, this determines weakness/resistance to attacks [referred to as the primary type]
- Type 2 = some Pokemon are dual type and have 2 [referred to as the secondary type]
- HP = hit points, or health, defines how much damage a Pokemon can withstand before fainting
- Attack =  the base modifier for normal attacks
- Defense = the base damage resistance against normal attacks
- Sp. Atk = special attack, the base modifier for special attacks
- Sp. Def = the base damage resistance against special attacks
- Speed = determines which Pokemon attacks first each round
- Generation = refers to which grouping/game series the Pokemon was released in
- Legendary = a boolean that identifies whether the Pokemon is legendary

## To check the rows and columns in a data

In [107]:
df.shape

(800, 12)

## To see all the columns in a data

In [108]:
df.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'HP', 'Attack', 'Defense', 'Sp. Atk',
       'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

## To check the datatypes of a column

In [109]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      359 non-null    object
 4   HP          800 non-null    int64 
 5   Attack      800 non-null    int64 
 6   Defense     800 non-null    int64 
 7   Sp. Atk     800 non-null    int64 
 8   Sp. Def     800 non-null    int64 
 9   Speed       800 non-null    int64 
 10  Generation  800 non-null    int64 
 11  Legendary   800 non-null    bool  
dtypes: bool(1), int64(8), object(3)
memory usage: 69.7+ KB


## statistical summary

In [110]:
df.describe().T # numeric cols

count       mean         std   min     25%    50%     75%    max
#           800.0  362.81375  208.343798   1.0  184.75  364.5  539.25  721.0
HP          800.0   69.25875   25.534669   1.0   50.00   65.0   80.00  255.0
Attack      800.0   79.00125   32.457366   5.0   55.00   75.0  100.00  190.0
Defense     800.0   73.84250   31.183501   5.0   50.00   70.0   90.00  230.0
Sp. Atk     800.0   72.82000   32.722294  10.0   49.75   65.0   95.00  194.0
Sp. Def     800.0   71.90250   27.828916  20.0   50.00   70.0   90.00  230.0
Speed       800.0   68.27750   29.060474   5.0   45.00   65.0   90.00  180.0
Generation  800.0    3.32375    1.661290   1.0    2.00    3.0    5.00    6.0

In [111]:
df.describe(include='object') # non numeric

Name Type 1  Type 2
count         800    800     359
unique        800     18      18
top     Bulbasaur  Water  Flying
freq            1    112      81

### from this we can say that type 2 has null values as it has 359 enteries.

top depicts : most frquent values
the top of name columns ca be any as it  has all unique enteries.



In [112]:
df["Type 2"].mode()

0    Flying
Name: Type 2, dtype: object

## To check the missing values in data

In [113]:
df.isnull().sum()

#               0
Name            0
Type 1          0
Type 2        441
HP              0
Attack          0
Defense         0
Sp. Atk         0
Sp. Def         0
Speed           0
Generation      0
Legendary       0
dtype: int64

## to see missing values %

In [114]:
df.isnull().sum()/df.shape[0]*100

#              0.000
Name           0.000
Type 1         0.000
Type 2        55.125
HP             0.000
Attack         0.000
Defense        0.000
Sp. Atk        0.000
Sp. Def        0.000
Speed          0.000
Generation     0.000
Legendary      0.000
dtype: float64

## To drop irrelavant columns : type 2 and #

In [115]:
df.drop(columns=['Type 2','#'],inplace=True)

In [116]:
df

Name   Type 1  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0                Bulbasaur    Grass  45      49       49       65       65   
1                  Ivysaur    Grass  60      62       63       80       80   
2                 Venusaur    Grass  80      82       83      100      100   
3    VenusaurMega Venusaur    Grass  80     100      123      122      120   
4               Charmander     Fire  39      52       43       60       50   
..                     ...      ...  ..     ...      ...      ...      ...   
795                Diancie     Rock  50     100      150      100      150   
796    DiancieMega Diancie     Rock  50     160      110      160      110   
797    HoopaHoopa Confined  Psychic  80     110       60      150      130   
798     HoopaHoopa Unbound  Psychic  80     160       60      170      130   
799              Volcanion     Fire  80     110      120      130       90   

     Speed  Generation  Legendary  
0       45           1      False  
1       60           1      False  
2       80           1      False  
3       80           1      False  
4       65           1      False  
..     ...         ...        ...  
795     50           6       True  
796    110           6       True  
797     70           6       True  
798     80           6       True  
799     70           6       True  

[800 rows x 10 columns]

## To read each column value

In [117]:
df['Name'] 

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

In [118]:
## to see all columns values using loop.
for i in df['Name']:
    
        
    print(i)
    


Bulbasaur
Ivysaur
Venusaur
VenusaurMega Venusaur
Charmander
Charmeleon
Charizard
CharizardMega Charizard X
CharizardMega Charizard Y
Squirtle
Wartortle
Blastoise
BlastoiseMega Blastoise
Caterpie
Metapod
Butterfree
Weedle
Kakuna
Beedrill
BeedrillMega Beedrill
Pidgey
Pidgeotto
Pidgeot
PidgeotMega Pidgeot
Rattata
Raticate
Spearow
Fearow
Ekans
Arbok
Pikachu
Raichu
Sandshrew
Sandslash
Nidoran (Female)
Nidorina
Nidoqueen
Nidoran (Male)
Nidorino
Nidoking
Clefairy
Clefable
Vulpix
Ninetales
Jigglypuff
Wigglytuff
Zubat
Golbat
Oddish
Gloom
Vileplume
Paras
Parasect
Venonat
Venomoth
Diglett
Dugtrio
Meowth
Persian
Psyduck
Golduck
Mankey
Primeape
Growlithe
Arcanine
Poliwag
Poliwhirl
Poliwrath
Abra
Kadabra
Alakazam
AlakazamMega Alakazam
Machop
Machoke
Machamp
Bellsprout
Weepinbell
Victreebel
Tentacool
Tentacruel
Geodude
Graveler
Golem
Ponyta
Rapidash
Slowpoke
Slowbro
SlowbroMega Slowbro
Magnemite
Magneton
Farfetch'd
Doduo
Dodrio
Seel
Dewgong
Grimer
Muk
Shellder
Cloyster
Gastly
Haunter
Gengar
GengarMeg

## To read multiple columns

In [19]:
df[['Name','Type 1','HP']]

Name   Type 1  HP
0                Bulbasaur    Grass  45
1                  Ivysaur    Grass  60
2                 Venusaur    Grass  80
3    VenusaurMega Venusaur    Grass  80
4               Charmander     Fire  39
..                     ...      ...  ..
795                Diancie     Rock  50
796    DiancieMega Diancie     Rock  50
797    HoopaHoopa Confined  Psychic  80
798     HoopaHoopa Unbound  Psychic  80
799              Volcanion     Fire  80

[800 rows x 3 columns]

## To read multiple columns with a certain rows

In [20]:
df[["Name",'HP']][400:411]

Name   HP
400                  Walrein  110
401                 Clamperl   35
402                  Huntail   55
403                 Gorebyss   55
404                Relicanth  100
405                  Luvdisc   43
406                    Bagon   45
407                  Shelgon   65
408                Salamence   95
409  SalamenceMega Salamence   95
410                   Beldum   40

## OR

In [21]:
df.iloc[400:411,[0,1]]

Name  Type 1
400                  Walrein     Ice
401                 Clamperl   Water
402                  Huntail   Water
403                 Gorebyss   Water
404                Relicanth   Water
405                  Luvdisc   Water
406                    Bagon  Dragon
407                  Shelgon  Dragon
408                Salamence  Dragon
409  SalamenceMega Salamence  Dragon
410                   Beldum   Steel

## To read particular  row

In [22]:
df.iloc[9]

Name          Squirtle
Type 1           Water
HP                  44
Attack              48
Defense             65
Sp. Atk             50
Sp. Def             64
Speed               43
Generation           1
Legendary        False
Name: 9, dtype: object

## To read multiple rows will all columns

In [23]:
df.iloc[10:16]

Name Type 1  HP  Attack  Defense  Sp. Atk  Sp. Def  \
10                Wartortle  Water  59      63       80       65       80   
11                Blastoise  Water  79      83      100       85      105   
12  BlastoiseMega Blastoise  Water  79     103      120      135      115   
13                 Caterpie    Bug  45      30       35       20       20   
14                  Metapod    Bug  50      20       55       25       25   
15               Butterfree    Bug  60      45       50       90       80   

    Speed  Generation  Legendary  
10     58           1      False  
11     78           1      False  
12     78           1      False  
13     45           1      False  
14     30           1      False  
15     70           1      False

In [24]:
df.iloc[10:16,2:5]

HP  Attack  Defense
10  59      63       80
11  79      83      100
12  79     103      120
13  45      30       35
14  50      20       55
15  60      45       50

## To read a specific name, i.e  Blastoise 

In [25]:
df.iloc[11,0]

'Blastoise'

In [26]:
# df.iloc[rows,column]
# df.iloc[30:46,0:2]
df.iloc[15,0][0]


'B'

## To iterate each rows of specific column

In [27]:
for i in df['Name']:
    print(i)

Bulbasaur
Ivysaur
Venusaur
VenusaurMega Venusaur
Charmander
Charmeleon
Charizard
CharizardMega Charizard X
CharizardMega Charizard Y
Squirtle
Wartortle
Blastoise
BlastoiseMega Blastoise
Caterpie
Metapod
Butterfree
Weedle
Kakuna
Beedrill
BeedrillMega Beedrill
Pidgey
Pidgeotto
Pidgeot
PidgeotMega Pidgeot
Rattata
Raticate
Spearow
Fearow
Ekans
Arbok
Pikachu
Raichu
Sandshrew
Sandslash
Nidoran (Female)
Nidorina
Nidoqueen
Nidoran (Male)
Nidorino
Nidoking
Clefairy
Clefable
Vulpix
Ninetales
Jigglypuff
Wigglytuff
Zubat
Golbat
Oddish
Gloom
Vileplume
Paras
Parasect
Venonat
Venomoth
Diglett
Dugtrio
Meowth
Persian
Psyduck
Golduck
Mankey
Primeape
Growlithe
Arcanine
Poliwag
Poliwhirl
Poliwrath
Abra
Kadabra
Alakazam
AlakazamMega Alakazam
Machop
Machoke
Machamp
Bellsprout
Weepinbell
Victreebel
Tentacool
Tentacruel
Geodude
Graveler
Golem
Ponyta
Rapidash
Slowpoke
Slowbro
SlowbroMega Slowbro
Magnemite
Magneton
Farfetch'd
Doduo
Dodrio
Seel
Dewgong
Grimer
Muk
Shellder
Cloyster
Gastly
Haunter
Gengar
GengarMeg

## To iterate multiple rows

In [28]:
for i,j in df.iterrows(): # here i is index and j is row data of each column
    print(i,"=----",j)      ## 
    #print(j[['Name','HP']])  ## fecthing specific columns 

0 =---- Name          Bulbasaur
Type 1            Grass
HP                   45
Attack               49
Defense              49
Sp. Atk              65
Sp. Def              65
Speed                45
Generation            1
Legendary         False
Name: 0, dtype: object
1 =---- Name          Ivysaur
Type 1          Grass
HP                 60
Attack             62
Defense            63
Sp. Atk            80
Sp. Def            80
Speed              60
Generation          1
Legendary       False
Name: 1, dtype: object
2 =---- Name          Venusaur
Type 1           Grass
HP                  80
Attack              82
Defense             83
Sp. Atk            100
Sp. Def            100
Speed               80
Generation           1
Legendary        False
Name: 2, dtype: object
3 =---- Name          VenusaurMega Venusaur
Type 1                        Grass
HP                               80
Attack                          100
Defense                         123
Sp. Atk                      

65 =---- Name          Poliwag
Type 1          Water
HP                 40
Attack             50
Defense            40
Sp. Atk            40
Sp. Def            40
Speed              90
Generation          1
Legendary       False
Name: 65, dtype: object
66 =---- Name          Poliwhirl
Type 1            Water
HP                   65
Attack               65
Defense              65
Sp. Atk              50
Sp. Def              50
Speed                90
Generation            1
Legendary         False
Name: 66, dtype: object
67 =---- Name          Poliwrath
Type 1            Water
HP                   90
Attack               95
Defense              95
Sp. Atk              70
Sp. Def              90
Speed                70
Generation            1
Legendary         False
Name: 67, dtype: object
68 =---- Name             Abra
Type 1        Psychic
HP                 25
Attack             20
Defense            15
Sp. Atk           105
Sp. Def            55
Speed              90
Generation      

150 =---- Name          Omastar
Type 1           Rock
HP                 70
Attack             60
Defense           125
Sp. Atk           115
Sp. Def            70
Speed              55
Generation          1
Legendary       False
Name: 150, dtype: object
151 =---- Name          Kabuto
Type 1          Rock
HP                30
Attack            80
Defense           90
Sp. Atk           55
Sp. Def           45
Speed             55
Generation         1
Legendary      False
Name: 151, dtype: object
152 =---- Name          Kabutops
Type 1            Rock
HP                  60
Attack             115
Defense            105
Sp. Atk             65
Sp. Def             70
Speed               80
Generation           1
Legendary        False
Name: 152, dtype: object
153 =---- Name          Aerodactyl
Type 1              Rock
HP                    80
Attack               105
Defense               65
Sp. Atk               60
Sp. Def               75
Speed                130
Generation             1


Name: 205, dtype: object
206 =---- Name          Sunkern
Type 1          Grass
HP                 30
Attack             30
Defense            30
Sp. Atk            30
Sp. Def            30
Speed              30
Generation          2
Legendary       False
Name: 206, dtype: object
207 =---- Name          Sunflora
Type 1           Grass
HP                  75
Attack              75
Defense             55
Sp. Atk            105
Sp. Def             85
Speed               30
Generation           2
Legendary        False
Name: 207, dtype: object
208 =---- Name          Yanma
Type 1          Bug
HP               65
Attack           65
Defense          45
Sp. Atk          75
Sp. Def          45
Speed            95
Generation        2
Legendary     False
Name: 208, dtype: object
209 =---- Name          Wooper
Type 1         Water
HP                55
Attack            45
Defense           45
Sp. Atk           25
Sp. Def           25
Speed             15
Generation         2
Legendary      False


Name: 291, dtype: object
292 =---- Name          Dustox
Type 1           Bug
HP                60
Attack            50
Defense           70
Sp. Atk           50
Sp. Def           90
Speed             65
Generation         3
Legendary      False
Name: 292, dtype: object
293 =---- Name          Lotad
Type 1        Water
HP               40
Attack           30
Defense          30
Sp. Atk          40
Sp. Def          50
Speed            30
Generation        3
Legendary     False
Name: 293, dtype: object
294 =---- Name          Lombre
Type 1         Water
HP                60
Attack            50
Defense           50
Sp. Atk           60
Sp. Def           70
Speed             50
Generation         3
Legendary      False
Name: 294, dtype: object
295 =---- Name          Ludicolo
Type 1           Water
HP                  80
Attack              70
Defense             70
Sp. Atk             90
Sp. Def            100
Speed               70
Generation           3
Legendary        False
Name: 295,

335 =---- Name          Medicham
Type 1        Fighting
HP                  60
Attack              60
Defense             75
Sp. Atk             60
Sp. Def             75
Speed               80
Generation           3
Legendary        False
Name: 335, dtype: object
336 =---- Name          MedichamMega Medicham
Type 1                     Fighting
HP                               60
Attack                          100
Defense                          85
Sp. Atk                          80
Sp. Def                          85
Speed                           100
Generation                        3
Legendary                     False
Name: 336, dtype: object
337 =---- Name          Electrike
Type 1         Electric
HP                   40
Attack               45
Defense              40
Sp. Atk              65
Sp. Def              40
Speed                65
Generation            3
Legendary         False
Name: 337, dtype: object
338 =---- Name          Manectric
Type 1         Electric
HP     

424 =---- Name          GroudonPrimal Groudon
Type 1                       Ground
HP                              100
Attack                          180
Defense                         160
Sp. Atk                         150
Sp. Def                          90
Speed                            90
Generation                        3
Legendary                      True
Name: 424, dtype: object
425 =---- Name          Rayquaza
Type 1          Dragon
HP                 105
Attack             150
Defense             90
Sp. Atk            150
Sp. Def             90
Speed               95
Generation           3
Legendary         True
Name: 425, dtype: object
426 =---- Name          RayquazaMega Rayquaza
Type 1                       Dragon
HP                              105
Attack                          180
Defense                         100
Sp. Atk                         180
Sp. Def                         100
Speed                           115
Generation                        3
Legend

Name: 460, dtype: object
461 =---- Name          Mothim
Type 1           Bug
HP                70
Attack            94
Defense           50
Sp. Atk           94
Sp. Def           50
Speed             66
Generation         4
Legendary      False
Name: 461, dtype: object
462 =---- Name          Combee
Type 1           Bug
HP                30
Attack            30
Defense           42
Sp. Atk           30
Sp. Def           42
Speed             70
Generation         4
Legendary      False
Name: 462, dtype: object
463 =---- Name          Vespiquen
Type 1              Bug
HP                   70
Attack               80
Defense             102
Sp. Atk              80
Sp. Def             102
Speed                40
Generation            4
Legendary         False
Name: 463, dtype: object
464 =---- Name          Pachirisu
Type 1         Electric
HP                   60
Attack               45
Defense              70
Sp. Atk              45
Sp. Def              90
Speed                95
Generati

Name: 566, dtype: object
567 =---- Name          Stoutland
Type 1           Normal
HP                   85
Attack              110
Defense              90
Sp. Atk              45
Sp. Def              90
Speed                80
Generation            5
Legendary         False
Name: 567, dtype: object
568 =---- Name          Purrloin
Type 1            Dark
HP                  41
Attack              50
Defense             37
Sp. Atk             50
Sp. Def             37
Speed               66
Generation           5
Legendary        False
Name: 568, dtype: object
569 =---- Name          Liepard
Type 1           Dark
HP                 64
Attack             88
Defense            50
Sp. Atk            88
Sp. Def            50
Speed             106
Generation          5
Legendary       False
Name: 569, dtype: object
570 =---- Name          Pansage
Type 1          Grass
HP                 50
Attack             53
Defense            48
Sp. Atk            53
Sp. Def            48
Speed           

Name: 600, dtype: object
601 =---- Name          Swadloon
Type 1             Bug
HP                  55
Attack              63
Defense             90
Sp. Atk             50
Sp. Def             80
Speed               42
Generation           5
Legendary        False
Name: 601, dtype: object
602 =---- Name          Leavanny
Type 1             Bug
HP                  75
Attack             103
Defense             80
Sp. Atk             70
Sp. Def             80
Speed               92
Generation           5
Legendary        False
Name: 602, dtype: object
603 =---- Name          Venipede
Type 1             Bug
HP                  30
Attack              45
Defense             59
Sp. Atk             30
Sp. Def             39
Speed               57
Generation           5
Legendary        False
Name: 603, dtype: object
604 =---- Name          Whirlipede
Type 1               Bug
HP                    40
Attack                55
Defense               99
Sp. Atk               40
Sp. Def             

Name: 690, dtype: object
691 =---- Name          Mandibuzz
Type 1             Dark
HP                  110
Attack               65
Defense             105
Sp. Atk              55
Sp. Def              95
Speed                80
Generation            5
Legendary         False
Name: 691, dtype: object
692 =---- Name          Heatmor
Type 1           Fire
HP                 85
Attack             97
Defense            66
Sp. Atk           105
Sp. Def            66
Speed              65
Generation          5
Legendary       False
Name: 692, dtype: object
693 =---- Name          Durant
Type 1           Bug
HP                58
Attack           109
Defense          112
Sp. Atk           48
Sp. Def           48
Speed            109
Generation         5
Legendary      False
Name: 693, dtype: object
694 =---- Name          Deino
Type 1         Dark
HP               52
Attack           65
Defense          50
Sp. Atk          45
Sp. Def          50
Speed            38
Generation        5
Legendary 

Name: 726, dtype: object
727 =---- Name          Bunnelby
Type 1          Normal
HP                  38
Attack              36
Defense             38
Sp. Atk             32
Sp. Def             36
Speed               57
Generation           6
Legendary        False
Name: 727, dtype: object
728 =---- Name          Diggersby
Type 1           Normal
HP                   85
Attack               56
Defense              77
Sp. Atk              50
Sp. Def              77
Speed                78
Generation            6
Legendary         False
Name: 728, dtype: object
729 =---- Name          Fletchling
Type 1            Normal
HP                    45
Attack                50
Defense               43
Sp. Atk               40
Sp. Def               38
Speed                 62
Generation             6
Legendary          False
Name: 729, dtype: object
730 =---- Name          Fletchinder
Type 1               Fire
HP                     62
Attack                 73
Defense                55
Sp. Atk   

In [29]:
for i,j in df.iterrows(): # here i is index and j is row data of each column
    print(j[['Name','HP']])  ## fecthing specific columns 

Name    Bulbasaur
HP             45
Name: 0, dtype: object
Name    Ivysaur
HP           60
Name: 1, dtype: object
Name    Venusaur
HP            80
Name: 2, dtype: object
Name    VenusaurMega Venusaur
HP                         80
Name: 3, dtype: object
Name    Charmander
HP              39
Name: 4, dtype: object
Name    Charmeleon
HP              58
Name: 5, dtype: object
Name    Charizard
HP             78
Name: 6, dtype: object
Name    CharizardMega Charizard X
HP                             78
Name: 7, dtype: object
Name    CharizardMega Charizard Y
HP                             78
Name: 8, dtype: object
Name    Squirtle
HP            44
Name: 9, dtype: object
Name    Wartortle
HP             59
Name: 10, dtype: object
Name    Blastoise
HP             79
Name: 11, dtype: object
Name    BlastoiseMega Blastoise
HP                           79
Name: 12, dtype: object
Name    Caterpie
HP            45
Name: 13, dtype: object
Name    Metapod
HP           50
Name: 14, dtype: object
Name

Name: 157, dtype: object
Name    Moltres
HP           90
Name: 158, dtype: object
Name    Dratini
HP           41
Name: 159, dtype: object
Name    Dragonair
HP             61
Name: 160, dtype: object
Name    Dragonite
HP             91
Name: 161, dtype: object
Name    Mewtwo
HP         106
Name: 162, dtype: object
Name    MewtwoMega Mewtwo X
HP                      106
Name: 163, dtype: object
Name    MewtwoMega Mewtwo Y
HP                      106
Name: 164, dtype: object
Name    Mew
HP      100
Name: 165, dtype: object
Name    Chikorita
HP             45
Name: 166, dtype: object
Name    Bayleef
HP           60
Name: 167, dtype: object
Name    Meganium
HP            80
Name: 168, dtype: object
Name    Cyndaquil
HP             39
Name: 169, dtype: object
Name    Quilava
HP           58
Name: 170, dtype: object
Name    Typhlosion
HP              78
Name: 171, dtype: object
Name    Totodile
HP            50
Name: 172, dtype: object
Name    Croconaw
HP            65
Name: 173, dtype: obje

Name    Breloom
HP           60
Name: 310, dtype: object
Name    Slakoth
HP           60
Name: 311, dtype: object
Name    Vigoroth
HP            80
Name: 312, dtype: object
Name    Slaking
HP          150
Name: 313, dtype: object
Name    Nincada
HP           31
Name: 314, dtype: object
Name    Ninjask
HP           61
Name: 315, dtype: object
Name    Shedinja
HP             1
Name: 316, dtype: object
Name    Whismur
HP           64
Name: 317, dtype: object
Name    Loudred
HP           84
Name: 318, dtype: object
Name    Exploud
HP          104
Name: 319, dtype: object
Name    Makuhita
HP            72
Name: 320, dtype: object
Name    Hariyama
HP           144
Name: 321, dtype: object
Name    Azurill
HP           50
Name: 322, dtype: object
Name    Nosepass
HP            30
Name: 323, dtype: object
Name    Skitty
HP          50
Name: 324, dtype: object
Name    Delcatty
HP            70
Name: 325, dtype: object
Name    Sableye
HP           50
Name: 326, dtype: object
Name    SableyeMega S

Name    Lumineon
HP            69
Name: 507, dtype: object
Name    Mantyke
HP           45
Name: 508, dtype: object
Name    Snover
HP          60
Name: 509, dtype: object
Name    Abomasnow
HP             90
Name: 510, dtype: object
Name    AbomasnowMega Abomasnow
HP                           90
Name: 511, dtype: object
Name    Weavile
HP           70
Name: 512, dtype: object
Name    Magnezone
HP             70
Name: 513, dtype: object
Name    Lickilicky
HP             110
Name: 514, dtype: object
Name    Rhyperior
HP            115
Name: 515, dtype: object
Name    Tangrowth
HP            100
Name: 516, dtype: object
Name    Electivire
HP              75
Name: 517, dtype: object
Name    Magmortar
HP             75
Name: 518, dtype: object
Name    Togekiss
HP            85
Name: 519, dtype: object
Name    Yanmega
HP           86
Name: 520, dtype: object
Name    Leafeon
HP           65
Name: 521, dtype: object
Name    Glaceon
HP           65
Name: 522, dtype: object
Name    Gliscor
HP    

Name: 665, dtype: object
Name    Elgyem
HP          55
Name: 666, dtype: object
Name    Beheeyem
HP            75
Name: 667, dtype: object
Name    Litwick
HP           50
Name: 668, dtype: object
Name    Lampent
HP           60
Name: 669, dtype: object
Name    Chandelure
HP              60
Name: 670, dtype: object
Name    Axew
HP        46
Name: 671, dtype: object
Name    Fraxure
HP           66
Name: 672, dtype: object
Name    Haxorus
HP           76
Name: 673, dtype: object
Name    Cubchoo
HP           55
Name: 674, dtype: object
Name    Beartic
HP           95
Name: 675, dtype: object
Name    Cryogonal
HP             70
Name: 676, dtype: object
Name    Shelmet
HP           50
Name: 677, dtype: object
Name    Accelgor
HP            80
Name: 678, dtype: object
Name    Stunfisk
HP           109
Name: 679, dtype: object
Name    Mienfoo
HP           45
Name: 680, dtype: object
Name    Mienshao
HP            65
Name: 681, dtype: object
Name    Druddigon
HP             77
Name: 682, dtype:

## To see the unique values in a particular column.

In [30]:
df['Type 1'].unique()
#print(df['Type 1'].unique())

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

## To see the Counts of the category in a data

In [31]:
df['Type 1'].value_counts()

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

In [32]:
df['Type 1'].mode()

0    Water
Name: Type 1, dtype: object

## To find the specific information in a data or to apply a condition in a data 

In [33]:

df[df['Type 1']=='Fire'] #it print only the data of Fire type1 pokemons

Name Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
4                   Charmander   Fire   39      52       43       60       50   
5                   Charmeleon   Fire   58      64       58       80       65   
6                    Charizard   Fire   78      84       78      109       85   
7    CharizardMega Charizard X   Fire   78     130      111      130       85   
8    CharizardMega Charizard Y   Fire   78     104       78      159      115   
42                      Vulpix   Fire   38      41       40       50       65   
43                   Ninetales   Fire   73      76       75       81      100   
63                   Growlithe   Fire   55      70       45       70       50   
64                    Arcanine   Fire   90     110       80      100       80   
83                      Ponyta   Fire   50      85       55       65       65   
84                    Rapidash   Fire   65     100       70       80       80   
135                     Magmar   Fire   65      95       57      100       85   
147                    Flareon   Fire   65     130       60       95      110   
158                    Moltres   Fire   90     100       90      125       85   
169                  Cyndaquil   Fire   39      52       43       60       50   
170                    Quilava   Fire   58      64       58       80       65   
171                 Typhlosion   Fire   78      84       78      109       85   
236                     Slugma   Fire   40      40       40       70       40   
237                   Magcargo   Fire   50      50      120       80       80   
259                      Magby   Fire   45      75       37       70       55   
263                      Entei   Fire  115     115       85       90       75   
270                      Ho-oh   Fire  106     130       90      110      154   
276                    Torchic   Fire   45      60       40       70       50   
277                  Combusken   Fire   60      85       60       85       60   
278                   Blaziken   Fire   80     120       70      110       70   
279      BlazikenMega Blaziken   Fire   80     160       80      130       80   
352                      Numel   Fire   60      60       40       65       45   
353                   Camerupt   Fire   70     100       70      105       75   
354      CameruptMega Camerupt   Fire   70     120      100      145      105   
355                    Torkoal   Fire   70      85      140       85       70   
435                   Chimchar   Fire   44      58       44       58       44   
436                   Monferno   Fire   64      78       52       78       52   
437                  Infernape   Fire   76     104       71      104       71   
518                  Magmortar   Fire   75      95       67      125       95   
542                    Heatran   Fire   91      90      106      130      106   
557                      Tepig   Fire   65      63       45       45       45   
558                    Pignite   Fire   90      93       55       70       55   
559                     Emboar   Fire  110     123       65      100       65   
572                    Pansear   Fire   50      53       48       53       48   
573                   Simisear   Fire   75      98       63       98       63   
614                   Darumaka   Fire   70      90       45       15       45   
615    DarmanitanStandard Mode   Fire  105     140       55       30       55   
616         DarmanitanZen Mode   Fire  105      30      105      140      105   
692                    Heatmor   Fire   85      97       66      105       66   
721                   Fennekin   Fire   40      45       40       62       60   
722                    Braixen   Fire   59      59       58       90       70   
723                    Delphox   Fire   75      69       72      114      100   
730                Fletchinder   Fire   62      73       55       56       52   
731                 Talonflame   Fire   78      81       

## it sort the values in ascending order

In [34]:
df.sort_values(by=['Name']) ## on the basis of Name

Name   Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
510                Abomasnow    Grass   90      92       75       92       85   
511  AbomasnowMega Abomasnow    Grass   90     132      105      132      105   
68                      Abra  Psychic   25      20       15      105       55   
392                    Absol     Dark   65     130       60       75       60   
393          AbsolMega Absol     Dark   65     150       60      115       60   
..                       ...      ...  ...     ...      ...      ...      ...   
632                  Zoroark     Dark   60     105       60      120       60   
631                    Zorua     Dark   40      65       40       80       40   
46                     Zubat   Poison   40      45       35       30       40   
695                 Zweilous     Dark   72      85       70       65       70   
794         Zygarde50% Forme   Dragon  108     100      121       81       95   

     Speed  Generation  Legendary  
510     60           4      False  
511     30           4      False  
68      90           1      False  
392     75           3      False  
393    115           3      False  
..     ...         ...        ...  
632    105           5      False  
631     65           5      False  
46      55           1      False  
695     58           5      False  
794     95           6       True  

[800 rows x 10 columns]

In [35]:
df.sort_values(by=['Attack'])

Name   Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
488                  Happiny   Normal  100       5        5       15       65   
121                  Chansey   Normal  250       5        5       35      105   
230                  Shuckle      Bug   20      10      230       10      230   
261                  Blissey   Normal  255      10       10       75      135   
139                 Magikarp    Water   20      10       55       15       20   
..                       ...      ...  ...     ...      ...      ...      ...   
429       DeoxysAttack Forme  Psychic   50     180       20      180       20   
426    RayquazaMega Rayquaza   Dragon  105     180      100      180      100   
424    GroudonPrimal Groudon   Ground  100     180      160      150       90   
232  HeracrossMega Heracross      Bug   80     185      115       40      105   
163      MewtwoMega Mewtwo X  Psychic  106     190      100      154      100   

     Speed  Generation  Legendary  
488     30           4      False  
121     50           1      False  
230      5           2      False  
261     55           2      False  
139     80           1      False  
..     ...         ...        ...  
429    150           3       True  
426    115           3       True  
424     90           3       True  
232     75           2      False  
163    130           1       True  

[800 rows x 10 columns]

In [36]:
df['Attack'].min()

5

## it sort the values in descending order

In [37]:
df.sort_values(by=['Name'],ascending=False) 

Name   Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
794         Zygarde50% Forme   Dragon  108     100      121       81       95   
695                 Zweilous     Dark   72      85       70       65       70   
46                     Zubat   Poison   40      45       35       30       40   
631                    Zorua     Dark   40      65       40       80       40   
632                  Zoroark     Dark   60     105       60      120       60   
..                       ...      ...  ...     ...      ...      ...      ...   
393          AbsolMega Absol     Dark   65     150       60      115       60   
392                    Absol     Dark   65     130       60       75       60   
68                      Abra  Psychic   25      20       15      105       55   
511  AbomasnowMega Abomasnow    Grass   90     132      105      132      105   
510                Abomasnow    Grass   90      92       75       92       85   

     Speed  Generation  Legendary  
794     95           6       True  
695     58           5      False  
46      55           1      False  
631     65           5      False  
632    105           5      False  
..     ...         ...        ...  
393    115           3      False  
392     75           3      False  
68      90           1      False  
511     30           4      False  
510     60           4      False  

[800 rows x 10 columns]

In [38]:
df.sort_values(['Attack'],ascending=False) 

Name   Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
163      MewtwoMega Mewtwo X  Psychic  106     190      100      154      100   
232  HeracrossMega Heracross      Bug   80     185      115       40      105   
424    GroudonPrimal Groudon   Ground  100     180      160      150       90   
426    RayquazaMega Rayquaza   Dragon  105     180      100      180      100   
429       DeoxysAttack Forme  Psychic   50     180       20      180       20   
..                       ...      ...  ...     ...      ...      ...      ...   
139                 Magikarp    Water   20      10       55       15       20   
261                  Blissey   Normal  255      10       10       75      135   
230                  Shuckle      Bug   20      10      230       10      230   
121                  Chansey   Normal  250       5        5       35      105   
488                  Happiny   Normal  100       5        5       15       65   

     Speed  Generation  Legendary  
163    130           1       True  
232     75           2      False  
424     90           3       True  
426    115           3       True  
429    150           3       True  
..     ...         ...        ...  
139     80           1      False  
261     55           2      False  
230      5           2      False  
121     50           1      False  
488     30           4      False  

[800 rows x 10 columns]

## when two columns name is given the it sort the values according to the first column

In [39]:
df.sort_values(["Attack","Name"]) 

Name   Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
121                  Chansey   Normal  250       5        5       35      105   
488                  Happiny   Normal  100       5        5       15       65   
261                  Blissey   Normal  255      10       10       75      135   
139                 Magikarp    Water   20      10       55       15       20   
230                  Shuckle      Bug   20      10      230       10      230   
..                       ...      ...  ...     ...      ...      ...      ...   
429       DeoxysAttack Forme  Psychic   50     180       20      180       20   
424    GroudonPrimal Groudon   Ground  100     180      160      150       90   
426    RayquazaMega Rayquaza   Dragon  105     180      100      180      100   
232  HeracrossMega Heracross      Bug   80     185      115       40      105   
163      MewtwoMega Mewtwo X  Psychic  106     190      100      154      100   

     Speed  Generation  Legendary  
121     50           1      False  
488     30           4      False  
261     55           2      False  
139     80           1      False  
230      5           2      False  
..     ...         ...        ...  
429    150           3       True  
424     90           3       True  
426    115           3       True  
232     75           2      False  
163    130           1       True  

[800 rows x 10 columns]

In [40]:
df['Type 1'].unique()

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

## here it shows both column in descending order¶

In [41]:
df.sort_values(["Type 1","Attack"],ascending=[False,False]) 

Name Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
141  GyaradosMega Gyarados  Water   95     155      109       70      130   
283  SwampertMega Swampert  Water  100     150      110       95      110   
422    KyogrePrimal Kyogre  Water  100     150       90      180      160   
349  SharpedoMega Sharpedo  Water   70     140       70      110       65   
107                Kingler  Water   55     130      115       50       50   
..                     ...    ...  ...     ...      ...      ...      ...   
446              Kricketot    Bug   37      25       41       25       41   
733                 Spewpa    Bug   45      22       60       27       30   
14                 Metapod    Bug   50      20       55       25       25   
179                 Ledyba    Bug   40      20       30       40       80   
230                Shuckle    Bug   20      10      230       10      230   

     Speed  Generation  Legendary  
141     81           1      False  
283     70           3      False  
422     90           3       True  
349    105           3      False  
107     75           1      False  
..     ...         ...        ...  
446     25           4      False  
733     29           6      False  
14      30           1      False  
179     55           2      False  
230      5           2      False  

[800 rows x 10 columns]

In [42]:
df["Attack"].max()

190

In [43]:
df[df["HP"]==df["HP"].max()]["Name"]

261    Blissey
Name: Name, dtype: object

## here it shows both column in ascending order¶

In [44]:
df.sort_values(["Attack","Type 1"]) 

Name   Type 1   HP  Attack  Defense  Sp. Atk  Sp. Def  \
121                  Chansey   Normal  250       5        5       35      105   
488                  Happiny   Normal  100       5        5       15       65   
230                  Shuckle      Bug   20      10      230       10      230   
261                  Blissey   Normal  255      10       10       75      135   
139                 Magikarp    Water   20      10       55       15       20   
..                       ...      ...  ...     ...      ...      ...      ...   
426    RayquazaMega Rayquaza   Dragon  105     180      100      180      100   
424    GroudonPrimal Groudon   Ground  100     180      160      150       90   
429       DeoxysAttack Forme  Psychic   50     180       20      180       20   
232  HeracrossMega Heracross      Bug   80     185      115       40      105   
163      MewtwoMega Mewtwo X  Psychic  106     190      100      154      100   

     Speed  Generation  Legendary  
121     50           1      False  
488     30           4      False  
230      5           2      False  
261     55           2      False  
139     80           1      False  
..     ...         ...        ...  
426    115           3       True  
424     90           3       True  
429    150           3       True  
232     75           2      False  
163    130           1       True  

[800 rows x 10 columns]

# Making changes to  the Data : Feature engineering:part of preprocessing)
mostly used in ML where ml model accurcay matters it works in cycle.

### Feature engineering:

### Feature engineering refers to manipulation — addition, deletion, combination, mutation — of your data set to improve machine learning model training, leading to better performance and greater accuracy. Effective feature engineering is based on sound knowledge of the business problem and the available data sources.


It involves:

- Feature creation
- feature transformation
- feature selection

 ### A common example of feature engineering is when your doctor uses your body mass index (BMI). BMI is calculated from both body weight and height, and serves as a surrogate for a characteristic that is very hard to accurately measure: the proportion of lean body mass.


in house price  data one column is no of bedrroms, no of washrooms  and these columns are important to predict price but we need to consider only one column so we do feature engineering by constructing  new feature square feet area from both in such a way no information gets lost.




Here , we compute total`and drop those individuals with out losing information

In [119]:
#Here i am added the columns HP, Attack, Defense, Sp.Atk, Sp.Def,Speed and named it as Total
#first input is going to colon which means all rows everything[:]
#then count the column here HP is the 2 column upto 9 column speed[:,2:8]
#then define the axis along which you want to add here axis=1 means horizontally


df['Total']=df.iloc[:,2:8].sum(axis=1)
df

Name   Type 1  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0                Bulbasaur    Grass  45      49       49       65       65   
1                  Ivysaur    Grass  60      62       63       80       80   
2                 Venusaur    Grass  80      82       83      100      100   
3    VenusaurMega Venusaur    Grass  80     100      123      122      120   
4               Charmander     Fire  39      52       43       60       50   
..                     ...      ...  ..     ...      ...      ...      ...   
795                Diancie     Rock  50     100      150      100      150   
796    DiancieMega Diancie     Rock  50     160      110      160      110   
797    HoopaHoopa Confined  Psychic  80     110       60      150      130   
798     HoopaHoopa Unbound  Psychic  80     160       60      170      130   
799              Volcanion     Fire  80     110      120      130       90   

     Speed  Generation  Legendary  Total  
0       45           1      False    318  
1       60           1      False    405  
2       80           1      False    525  
3       80           1      False    625  
4       65           1      False    309  
..     ...         ...        ...    ...  
795     50           6       True    600  
796    110           6       True    700  
797     70           6       True    600  
798     80           6       True    680  
799     70           6       True    600  

[800 rows x 11 columns]

 ## Total sum of all stats that come after this, a general guide to how strong a Pokemon is

In [120]:
df["Total"].max()  ## strongest pokemon stats


780

## Who is strongest pokemon

In [121]:
df[df['Total']==780]["Name"]

163      MewtwoMega Mewtwo X
164      MewtwoMega Mewtwo Y
426    RayquazaMega Rayquaza
Name: Name, dtype: object

## Arrange the column order 

In [48]:
col=list(df.columns)
col

['Name',
 'Type 1',
 'HP',
 'Attack',
 'Defense',
 'Sp. Atk',
 'Sp. Def',
 'Speed',
 'Generation',
 'Legendary',
 'Total']

In [49]:
col[0:2]+[col[-1]]+col[2:10]  ## concatenation

['Name',
 'Type 1',
 'Total',
 'HP',
 'Attack',
 'Defense',
 'Sp. Atk',
 'Sp. Def',
 'Speed',
 'Generation',
 'Legendary']

In [50]:
# Arranging the columns

df=df[col[0:2]+[col[-1]]+col[2:10]]
df


Name   Type 1  Total  HP  Attack  Defense  Sp. Atk  \
0                Bulbasaur    Grass    318  45      49       49       65   
1                  Ivysaur    Grass    405  60      62       63       80   
2                 Venusaur    Grass    525  80      82       83      100   
3    VenusaurMega Venusaur    Grass    625  80     100      123      122   
4               Charmander     Fire    309  39      52       43       60   
..                     ...      ...    ...  ..     ...      ...      ...   
795                Diancie     Rock    600  50     100      150      100   
796    DiancieMega Diancie     Rock    700  50     160      110      160   
797    HoopaHoopa Confined  Psychic    600  80     110       60      150   
798     HoopaHoopa Unbound  Psychic    680  80     160       60      170   
799              Volcanion     Fire    600  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
4         50     65           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[800 rows x 11 columns]

## Drop

In [51]:
pokemon=df.drop(columns=['HP','Attack','Defense','Sp. Atk','Sp. Def','Speed'])

In [52]:
pokemon

Name   Type 1  Total  Generation  Legendary
0                Bulbasaur    Grass    318           1      False
1                  Ivysaur    Grass    405           1      False
2                 Venusaur    Grass    525           1      False
3    VenusaurMega Venusaur    Grass    625           1      False
4               Charmander     Fire    309           1      False
..                     ...      ...    ...         ...        ...
795                Diancie     Rock    600           6       True
796    DiancieMega Diancie     Rock    700           6       True
797    HoopaHoopa Confined  Psychic    600           6       True
798     HoopaHoopa Unbound  Psychic    680           6       True
799              Volcanion     Fire    600           6       True

[800 rows x 5 columns]

## Saving our transformed/cleaned Data (Exporting into desired format)

## To save the dataset

In [53]:
pokemon.to_csv('updated_pokemon.csv',index=False) #here index=False to delete the extra indexes of csv format


In [54]:
# pokemon

In [55]:
##pokemon=pd.read_csv('updated_pokemon.csv')

## Filtering of data Or Condition in data

In [56]:
df["Type 1"].unique()

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

In [57]:
df[(df['Type 1']=="Grass") | (df['Legendary']==True)]

Name   Type 1  Total  HP  Attack  Defense  Sp. Atk  \
0                Bulbasaur    Grass    318  45      49       49       65   
1                  Ivysaur    Grass    405  60      62       63       80   
2                 Venusaur    Grass    525  80      82       83      100   
3    VenusaurMega Venusaur    Grass    625  80     100      123      122   
48                  Oddish    Grass    320  45      50       55       75   
..                     ...      ...    ...  ..     ...      ...      ...   
795                Diancie     Rock    600  50     100      150      100   
796    DiancieMega Diancie     Rock    700  50     160      110      160   
797    HoopaHoopa Confined  Psychic    600  80     110       60      150   
798     HoopaHoopa Unbound  Psychic    680  80     160       60      170   
799              Volcanion     Fire    600  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
48        65     30           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[132 rows x 11 columns]

In [58]:
df[(df['Attack']==df['Attack'].min()) | (df['HP']==df['HP'].max())]

Name  Type 1  Total   HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
121  Chansey  Normal    450  250       5        5       35      105     50   
261  Blissey  Normal    540  255      10       10       75      135     55   
488  Happiny  Normal    220  100       5        5       15       65     30   

     Generation  Legendary  
121           1      False  
261           2      False  
488           4      False

In [59]:
## or

In [60]:
df.iloc[np.where((df['Attack']==min(df['Attack'])) | (df['HP']==max(df['HP'])))]


Name  Type 1  Total   HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
121  Chansey  Normal    450  250       5        5       35      105     50   
261  Blissey  Normal    540  255      10       10       75      135     55   
488  Happiny  Normal    220  100       5        5       15       65     30   

     Generation  Legendary  
121           1      False  
261           2      False  
488           4      False

In [61]:
# df[(df['Type 1']=="Dragon") & (df['Type 2']=='Fire')]

In [62]:
df[(df['Type 1']=="Grass") | (df['Legendary']==True)]

Name   Type 1  Total  HP  Attack  Defense  Sp. Atk  \
0                Bulbasaur    Grass    318  45      49       49       65   
1                  Ivysaur    Grass    405  60      62       63       80   
2                 Venusaur    Grass    525  80      82       83      100   
3    VenusaurMega Venusaur    Grass    625  80     100      123      122   
48                  Oddish    Grass    320  45      50       55       75   
..                     ...      ...    ...  ..     ...      ...      ...   
795                Diancie     Rock    600  50     100      150      100   
796    DiancieMega Diancie     Rock    700  50     160      110      160   
797    HoopaHoopa Confined  Psychic    600  80     110       60      150   
798     HoopaHoopa Unbound  Psychic    680  80     160       60      170   
799              Volcanion     Fire    600  80     110      120      130   

     Sp. Def  Speed  Generation  Legendary  
0         65     45           1      False  
1         80     60           1      False  
2        100     80           1      False  
3        120     80           1      False  
48        65     30           1      False  
..       ...    ...         ...        ...  
795      150     50           6       True  
796      110    110           6       True  
797      130     70           6       True  
798      130     80           6       True  
799       90     70           6       True  

[132 rows x 11 columns]

In [63]:
df[(df['Type 1']=='Grass') & (df['Generation']==6) & (df['Attack']>100)]

Name Type 1  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
720  Chesnaught  Grass    530  88     107      122       74       75     64   

     Generation  Legendary  
720           6      False

In [64]:
df[df['Name']=='Pikachu']#To find the any name in the data

Name    Type 1  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
30  Pikachu  Electric    320  35      55       40       50       50     90   

    Generation  Legendary  
30           1      False

In [65]:
df[df['Name']=='Raichu']

Name    Type 1  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
31  Raichu  Electric    485  60      90       55       90       80    110   

    Generation  Legendary  
31           1      False

In [66]:
# Here loc is for locating a string based column
#df.loc[df['Type 1']=='Fire','Type 1']='Flamer' #it chnages the Fire to Flamer in Type 1 column
#df

In [67]:
#df[df['Type 1']=='Flamer']

In [68]:
#df[(df['Type 1']=='Flamer') & (df['Generation']==6) & (df['Legendary']==True)]

### Insights 

In [69]:
df.dtypes

Name          object
Type 1        object
Total          int64
HP             int64
Attack         int64
Defense        int64
Sp. Atk        int64
Sp. Def        int64
Speed          int64
Generation     int64
Legendary       bool
dtype: object

## How many types of pokemon exist in this catelog

In [70]:
len(df["Type 1"].unique())

18

####  Insight : 18 types of pokemon exist 

## Following are the categories 

In [71]:
df["Type 1"].unique()

array(['Grass', 'Fire', 'Water', 'Bug', 'Normal', 'Poison', 'Electric',
       'Ground', 'Fairy', 'Fighting', 'Psychic', 'Rock', 'Ghost', 'Ice',
       'Dragon', 'Dark', 'Steel', 'Flying'], dtype=object)

## Which pokemon type has highest and lowest instances

In [72]:
df["Type 1"].value_counts()

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type 1, dtype: int64

####  water type pokemon is more and flying pokemon is less

### How many legends pokemon 

In [73]:
df["Legendary"].unique()

array([False,  True])

In [74]:
df["Legendary"].value_counts()

False    735
True      65
Name: Legendary, dtype: int64

## Upto which generation pokemon exits 

In [75]:
df["Generation"].unique()

array([1, 2, 3, 4, 5, 6], dtype=int64)

## Which generation has more pokemon

In [76]:
df["Generation"].value_counts()

1    166
5    165
3    160
4    121
2    106
6     82
Name: Generation, dtype: int64

##  Which category of  pokemon  is strongest and weekest

In [77]:
df[df["Total"]==df["Total"].max()]["Type 1"]

163    Psychic
164    Psychic
426     Dragon
Name: Type 1, dtype: object

In [78]:
df[df["Total"]==df["Total"].min()]["Type 1"]

206    Grass
Name: Type 1, dtype: object

In [79]:
#df.sort_values(by="Total",ascending=False)

In [80]:
#df.sort_values(by="Total",ascending=False).tail()  ## weekest

### Strongest pokemon of types : Dragon and Psychic type
### weekest :  Grass type

## Find top 5 pokemon based on total 

In [81]:
df.sort_values(by="Total",ascending=False).head(5)

Name   Type 1  Total   HP  Attack  Defense  Sp. Atk  \
426  RayquazaMega Rayquaza   Dragon    780  105     180      100      180   
164    MewtwoMega Mewtwo Y  Psychic    780  106     150       70      194   
163    MewtwoMega Mewtwo X  Psychic    780  106     190      100      154   
422    KyogrePrimal Kyogre    Water    770  100     150       90      180   
424  GroudonPrimal Groudon   Ground    770  100     180      160      150   

     Sp. Def  Speed  Generation  Legendary  
426      100    115           3       True  
164      120    140           1       True  
163      100    130           1       True  
422      160     90           3       True  
424       90     90           3       True

## Find the maximum,min and average stats of each category of pokemon

In [82]:
df.groupby(['Type 1'])['Total'].agg(["max","mean","min"])


max        mean  min
Type 1                        
Bug       600  378.927536  194
Dark      680  445.741935  220
Dragon    780  550.531250  300
Electric  610  443.409091  205
Fairy     680  413.176471  218
Fighting  625  416.444444  210
Fire      680  458.076923  250
Flying    580  485.000000  245
Ghost     680  439.562500  275
Grass     630  421.142857  180
Ground    770  437.500000  265
Ice       580  433.458333  250
Normal    720  401.683673  190
Poison    535  399.142857  245
Psychic   780  475.947368  198
Rock      700  453.750000  280
Steel     700  487.703704  300
Water     770  430.455357  200

In [83]:
x=df.groupby(['Type 1'])['Total'].agg(["max","mean","min"])
x.sort_values(by="mean",ascending=False)
## or df.groupby(['Type 1'])['Total'].agg(["max","mean","min"]).sort_values(by="mean")

max        mean  min
Type 1                        
Dragon    780  550.531250  300
Steel     700  487.703704  300
Flying    580  485.000000  245
Psychic   780  475.947368  198
Fire      680  458.076923  250
Rock      700  453.750000  280
Dark      680  445.741935  220
Electric  610  443.409091  205
Ghost     680  439.562500  275
Ground    770  437.500000  265
Ice       580  433.458333  250
Water     770  430.455357  200
Grass     630  421.142857  180
Fighting  625  416.444444  210
Fairy     680  413.176471  218
Normal    720  401.683673  190
Poison    535  399.142857  245
Bug       600  378.927536  194

## Dragon type of pokemon is powerful  and Bug is weekest based on average stats points,

### details of Dragon type pokemon

In [84]:
df[df["Type 1"]=="Dragon"]

Name  Type 1  Total   HP  Attack  Defense  Sp. Atk  \
159                  Dratini  Dragon    300   41      64       45       50   
160                Dragonair  Dragon    420   61      84       65       70   
161                Dragonite  Dragon    600   91     134       95      100   
365                  Altaria  Dragon    490   75      70       90       70   
366      AltariaMega Altaria  Dragon    590   75     110      110      110   
406                    Bagon  Dragon    300   45      75       60       40   
407                  Shelgon  Dragon    420   65      95      100       60   
408                Salamence  Dragon    600   95     135       80      110   
409  SalamenceMega Salamence  Dragon    700   95     145      130      120   
417                   Latias  Dragon    600   80      80       90      110   
418        LatiasMega Latias  Dragon    700   80     100      120      140   
419                   Latios  Dragon    600   80      90       80      130   
420        LatiosMega Latios  Dragon    700   80     130      100      160   
425                 Rayquaza  Dragon    680  105     150       90      150   
426    RayquazaMega Rayquaza  Dragon    780  105     180      100      180   
491                    Gible  Dragon    300   58      70       45       40   
492                   Gabite  Dragon    410   68      90       65       50   
493                 Garchomp  Dragon    600  108     130       95       80   
494    GarchompMega Garchomp  Dragon    700  108     170      115      120   
671                     Axew  Dragon    320   46      87       60       30   
672                  Fraxure  Dragon    410   66     117       70       40   
673                  Haxorus  Dragon    540   76     147       90       60   
682                Druddigon  Dragon    485   77     120       90       60   
706                 Reshiram  Dragon    680  100     120      100      150   
707                   Zekrom  Dragon    680  100     150      120      120   
710                   Kyurem  Dragon    660  125     130       90      130   
711       KyuremBlack Kyurem  Dragon    700  125     170      100      120   
712       KyuremWhite Kyurem  Dragon    700  125     120       90      170   
774                    Goomy  Dragon    300   45      50       35       55   
775                  Sliggoo  Dragon    452   68      75       53       83   
776                   Goodra  Dragon    600   90     100       70      110   
794         Zygarde50% Forme  Dragon    600  108     100      121       81   

     Sp. Def  Speed  Generation  Legendary  
159       50     50           1      False  
160       70     70           1      False  
161      100     80           1      False  
365      105     80           3      False  
366      105     80           3      False  
406       30     50           3      False  
407       50     50           3      False  
408       80    100           3      False  
409       90    120           3      False  
417      130    110           3       True  
418      150    110           3       True  
419      110    110           3       True  
420      120    110           3       True  
425       90     95           3       True  
426      100    115           3       True  
491       45     42           4      False  
492       55     82           4      False  
493       85    102           4      False  
494       95     92           4      False  
671       40     57           5      False  
672       50     67           5      False  
673       70     97           5      False  
682       90     48           5      False  
706      120     90           5       True  
707      100     90           5       True  
710       90     95           5       True  
711       90     95           5       True  
712      100     95           5       True  
774       75     40           6      False  
775      113     60           6      False  
776      150     80           6      False  
794      

In [85]:
df[(df["Type 1"]=="Dragon") & (df["Total"]==780)]

Name  Type 1  Total   HP  Attack  Defense  Sp. Atk  \
426  RayquazaMega Rayquaza  Dragon    780  105     180      100      180   

     Sp. Def  Speed  Generation  Legendary  
426      100    115           3       True

## Which pokeman whose total points is max and min

In [86]:
df[df['Total']==df['Total'].max()]

Name   Type 1  Total   HP  Attack  Defense  Sp. Atk  \
163    MewtwoMega Mewtwo X  Psychic    780  106     190      100      154   
164    MewtwoMega Mewtwo Y  Psychic    780  106     150       70      194   
426  RayquazaMega Rayquaza   Dragon    780  105     180      100      180   

     Sp. Def  Speed  Generation  Legendary  
163      100    130           1       True  
164      120    140           1       True  
426      100    115           3       True

In [87]:
df[df['Total']==df['Total'].min()]

Name Type 1  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
206  Sunkern  Grass    180  30      30       30       30       30     30   

     Generation  Legendary  
206           2      False

## Find the average,max and min total stats of each type 1 along with pokemon names

In [88]:
df.groupby(['Type 1','Name'])['Total'].agg(['mean','max','min'])

mean  max  min
Type 1 Name                                  
Bug    Accelgor               495.0  495  495
       Ariados                390.0  390  390
       Beautifly              395.0  395  395
       Beedrill               395.0  395  395
       BeedrillMega Beedrill  495.0  495  495
...                             ...  ...  ...
Water  Wailord                500.0  500  500
       Wartortle              405.0  405  405
       Whiscash               468.0  468  468
       Wingull                270.0  270  270
       Wooper                 210.0  210  210

[800 rows x 3 columns]

### Which generation' pokemon is strongest

In [89]:
df.groupby(['Generation'])['Total'].agg(['mean','max','min']).sort_values(by="max",ascending=False)

mean  max  min
Generation                      
1           426.813253  780  195
3           436.225000  780  190
4           459.016529  720  194
2           418.283019  700  180
5           434.987879  700  255
6           436.378049  700  200

### of Which generation which type of pokemon is strongest

In [90]:
df.groupby(['Generation','Type 1'])['Total'].agg(['mean','max','min'])

mean  max  min
Generation Type 1                        
1          Bug       366.428571  600  195
           Dragon    440.000000  600  300
           Electric  444.444444  580  320
           Fairy     403.000000  483  323
           Fighting  412.142857  505  305
...                         ...  ...  ...
6          Poison    407.000000  494  320
           Psychic   513.400000  680  355
           Rock      485.777778  700  306
           Steel     456.600000  520  325
           Water     415.800000  530  314

[98 rows x 3 columns]

## Top 5 pokemon based on hp

In [91]:
df.sort_values(by="HP",ascending=False)[["Name","Type 1","Generation","Legendary","HP"]].head(5)

Name   Type 1  Generation  Legendary   HP
261    Blissey   Normal           2      False  255
121    Chansey   Normal           1      False  250
217  Wobbuffet  Psychic           2      False  190
351    Wailord    Water           3      False  170
655  Alomomola    Water           5      False  165

## Normal type has highest HP


## Find max,min and average HP  of each type

In [92]:
df.groupby(["Type 1"])["HP"].agg(["max","min","mean"]).sort_values(by="max",ascending=False)

max  min       mean
Type 1                       
Normal    255   30  77.275510
Psychic   190   20  70.631579
Water     170   20  72.062500
Ghost     150   20  64.437500
Fighting  144   30  69.851852
Dark      126   35  66.806452
Fairy     126   35  74.117647
Dragon    125   41  83.312500
Rock      123   30  65.363636
Grass     123   30  67.271429
Ground    115   10  73.781250
Fire      115   38  69.903846
Ice       110   36  72.000000
Poison    105   35  67.250000
Steel     100   40  65.222222
Electric   90   20  59.795455
Bug        86    1  56.884058
Flying     85   40  70.750000

### normal type has highest hp
### flyinh type has lowest hp

## Top 5 pokemon based on Attack

In [93]:
df.sort_values(by="Attack",ascending=False)[["Name","Type 1","Generation","Legendary","Attack"]].head(5)

Name   Type 1  Generation  Legendary  Attack
163      MewtwoMega Mewtwo X  Psychic           1       True     190
232  HeracrossMega Heracross      Bug           2      False     185
424    GroudonPrimal Groudon   Ground           3       True     180
426    RayquazaMega Rayquaza   Dragon           3       True     180
429       DeoxysAttack Forme  Psychic           3       True     180

## Find max,min and average attack of each Type

In [94]:
df.groupby(["Type 1"])["Attack"].agg(["max","min","mean"]).sort_values(by="max",ascending=False)

max  min        mean
Type 1                        
Psychic   190   20   71.456140
Bug       185   10   70.971014
Dragon    180   50  112.125000
Ground    180   40   95.750000
Ghost     165   30   73.781250
Rock      165   40   92.863636
Normal    160    5   73.469388
Fire      160   30   84.769231
Water     155   10   74.151786
Dark      150   50   88.387097
Steel     150   24   92.703704
Fighting  145   35   96.777778
Grass     132   27   73.214286
Fairy     131   20   61.529412
Ice       130   30   72.750000
Electric  123   30   69.090909
Flying    115   30   78.750000
Poison    106   43   74.678571

## psycic has strongest attck
## Poison has weekest attack.

## Top 5 pokemon based on Speed

In [95]:
df.sort_values(by="Speed",ascending=False)[["Name","Type 1","Generation","Legendary","Speed"]].head(5)

Name   Type 1  Generation  Legendary  Speed
431          DeoxysSpeed Forme  Psychic           3       True    180
315                    Ninjask      Bug           3      False    160
428         DeoxysNormal Forme  Psychic           3       True    150
154  AerodactylMega Aerodactyl     Rock           1      False    150
71       AlakazamMega Alakazam  Psychic           1      False    150

## Findmax,min and average speed of each type

In [96]:
df.groupby(["Type 1"])["Speed"].agg(["max","min","mean"]).sort_values(by="max",ascending=False)

max  min        mean
Type 1                        
Psychic   180   20   81.491228
Bug       160    5   61.681159
Rock      150   10   55.909091
Grass     145   10   61.928571
Electric  140   35   84.500000
Normal    135    5   71.551020
Ghost     130   20   64.343750
Poison    130   25   63.571429
Fire      126   20   74.442308
Dark      125   20   76.161290
Flying    123   55  102.500000
Water     122   15   65.964286
Ground    120   10   63.906250
Dragon    120   40   83.031250
Fighting  118   25   66.074074
Ice       110   25   63.458333
Steel     110   23   55.259259
Fairy      99   15   48.588235

## psycic has high spped
## fairy has low speed

## Top 5 pokemon based on Defence

In [97]:
df.sort_values(by="Defense",ascending=False)[["Name","Type 1","Generation","Legendary","Defense"]].head(5)

Name Type 1  Generation  Legendary  Defense
230              Shuckle    Bug           2      False      230
224  SteelixMega Steelix  Steel           2      False      230
333    AggronMega Aggron  Steel           3      False      230
223              Steelix  Steel           2      False      200
414             Regirock   Rock           3       True      200

## Find max,min and average defense of each type

In [98]:
df.groupby(["Type 1"])["Defense"].agg(["max","min","mean"]).sort_values(by="max",ascending=False)

max  min        mean
Type 1                        
Bug       230   30   70.724638
Steel     230   50  126.370370
Rock      200   40  100.795455
Ice       184   15   71.416667
Water     180   20   72.946429
Ground    160   25   84.843750
Psychic   160   15   67.684211
Ghost     145   30   81.187500
Fire      140   37   67.769231
Grass     131   30   70.800000
Dragon    130   35   86.375000
Normal    126    5   59.846939
Dark      125   30   70.225806
Poison    120   35   68.821429
Electric  115   15   66.295455
Fighting   95   30   65.925926
Fairy      95   28   65.705882
Flying     80   35   66.250000

## bug has high defense power
## flying has low defense power

## Top 5 pokemon based on spl. attack.

In [99]:
df.sort_values(by="Sp. Atk",ascending=False)[["Name","Type 1","Generation","Legendary","Sp. Atk"]].head(5)

Name   Type 1  Generation  Legendary  Sp. Atk
164    MewtwoMega Mewtwo Y  Psychic           1       True      194
422    KyogrePrimal Kyogre    Water           3       True      180
429     DeoxysAttack Forme  Psychic           3       True      180
426  RayquazaMega Rayquaza   Dragon           3       True      180
71   AlakazamMega Alakazam  Psychic           1      False      175

## Find max,min and average spl.attack of each type

In [100]:
df.groupby(["Type 1"])["Sp. Atk"].agg(["max","min","mean"]).sort_values(by="mean",ascending=False)

max  min       mean
Type 1                       
Psychic   194   23  98.403509
Dragon    180   30  96.843750
Flying    125   45  94.250000
Electric  165   35  90.022727
Fire      159   15  88.980769
Ghost     170   30  79.343750
Fairy     131   40  78.529412
Ice       130   30  77.541667
Grass     145   24  77.500000
Water     180   10  74.812500
Dark      140   30  74.645161
Steel     150   24  67.518519
Rock      160   10  63.340909
Poison    100   30  60.428571
Ground    150   20  56.468750
Normal    135   15  55.816327
Bug       135   10  53.869565
Fighting  140   20  53.111111

## Top 5 pokemon based on spl. defence.

In [101]:
df.sort_values(by="Sp. Def",ascending=False)[["Name","Type 1","Generation","Legendary","Sp. Def"]].head(5)

Name   Type 1  Generation  Legendary  Sp. Def
230              Shuckle      Bug           2      False      230
415               Regice      Ice           3       True      200
430  DeoxysDefense Forme  Psychic           3       True      160
422  KyogrePrimal Kyogre    Water           3       True      160
270                Ho-oh     Fire           2       True      154

## Find max,min and average spl.defense of each type

In [102]:
df.groupby(["Type 1"])["Sp. Def"].agg(["max","min","mean"]).sort_values(by="mean",ascending=False)

max  min       mean
Type 1                       
Dragon    150   30  88.843750
Psychic   160   20  86.280702
Fairy     154   40  84.705882
Steel     150   37  80.629630
Ghost     135   33  76.468750
Ice       200   30  76.291667
Rock      150   25  75.477273
Electric  110   32  73.704545
Flying     90   40  72.500000
Fire      154   40  72.211538
Water     160   20  70.517857
Grass     129   30  70.428571
Dark      130   30  69.516129
Bug       230   20  64.797101
Fighting  110   30  64.703704
Poison    123   40  64.392857
Normal    135   20  63.724490
Ground    120   30  62.750000